File: tools/ghidra/legacy/NativeI2CStaticReport.java
// @category Aorus

import java.util.*;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.data.DataType;
import ghidra.program.model.data.StringDataInstance;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.symbol.ExternalLocation;
import ghidra.program.model.symbol.ExternalManager;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceManager;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;
import ghidra.program.model.symbol.SymbolTable;
import ghidra.util.task.ConsoleTaskMonitor;

public class NativeI2CStaticReport extends GhidraScript {
    private final String[] patterns = {
        "I2C", "DDC", "NVAPI", "NvAPI", "Boost", "Intel", "ctlI2CAccess",
        "GvIntelI2C", "GvWriteI2C", "GvReadI2C", "nvapi_Direct", "Direct_Get"
    };

    private final String[] targets = {
        "GvWriteI2C", "GvReadI2C", "GvInitDispLib", "GvFreeDispLib",
        "GvIsNvBoost5", "nvapi_Direct_Get", "nvapi_Direct_GetMethod"
    };

    @Override
    public void run() throws Exception {
        println("# PROGRAM " + currentProgram.getName());
        println("IMAGE_BASE " + currentProgram.getImageBase());
        println("LANGUAGE " + currentProgram.getLanguageID());
        println("COMPILER " + currentProgram.getCompilerSpec().getCompilerSpecID());
        println("");
        printMemory();
        printExports();
        printImports();
        printStrings();
        printTargetFunctions();
    }

    private void printMemory() {
        println("## MEMORY BLOCKS");
        for (MemoryBlock b : currentProgram.getMemory().getBlocks()) {
            println(b.getName() + " " + b.getStart() + "-" + b.getEnd() +
                " R" + b.isRead() + " W" + b.isWrite() + " X" + b.isExecute());
        }
        println("");
    }

    private void printExports() {
        println("## EXPORTS");
        SymbolTable st = currentProgram.getSymbolTable();
        SymbolIterator it = st.getAllSymbols(true);
        while (it.hasNext()) {
            Symbol s = it.next();
            if (st.isExternalEntryPoint(s.getAddress())) {
                println(s.getAddress() + " " + s.getName(true));
            }
        }
        println("");
    }

    private void printImports() {
        println("## IMPORTS/EXTERNALS");
        ExternalManager em = currentProgram.getExternalManager();
        for (String lib : em.getExternalLibraryNames()) {
            println("LIB " + lib);
            Iterator<ExternalLocation> it = em.getExternalLocations(lib);
            while (it.hasNext()) {
                ExternalLocation loc = it.next();
                println("  " + loc.getAddress() + " " + loc.getLabel());
            }
        }
        println("");
    }

    private void printStrings() {
        println("## MATCHED STRINGS");
        Iterator<Data> it = currentProgram.getListing().getDefinedData(true);
        while (it.hasNext()) {
            Data d = it.next();
            String value = null;
            try {
                StringDataInstance sdi = StringDataInstance.getStringDataInstance(d);
                if (sdi != null) {
                    value = sdi.getStringValue();
                }
            }
            catch (Exception e) {
                value = null;
            }
            if (value == null) {
                Object raw = d.getValue();
                if (raw instanceof String) {
                    value = (String) raw;
                }
            }
            if (value == null) {
                continue;
            }
            String lower = value.toLowerCase(Locale.ROOT);
            for (String p : patterns) {
                if (lower.contains(p.toLowerCase(Locale.ROOT))) {
                    println(d.getAddress() + " " + value.replace("\n", "\\n"));
                    break;
                }
            }
        }
        println("");
    }

    private void printTargetFunctions() throws Exception {
        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);
        println("## TARGET FUNCTIONS");
        for (String name : targets) {
            List<Function> funcs = findFunctionsByName(name);
            for (Function f : funcs) {
                println("### " + f.getName() + " " + f.getEntryPoint());
                println("SIGNATURE: " + f.getSignature());
                println("CALLS: " + String.join(", ", getCalledNames(f)));
                DecompileResults res = ifc.decompileFunction(f, 90, new ConsoleTaskMonitor());
                if (res != null && res.decompileCompleted()) {
                    println(res.getDecompiledFunction().getC());
                }
                else {
                    println("<decompile failed>");
                }
                println("");
            }
        }
        println("");
    }

    private List<Function> findFunctionsByName(String name) {
        ArrayList<Function> funcs = new ArrayList<>();
        FunctionIterator it = currentProgram.getFunctionManager().getFunctions(true);
        while (it.hasNext()) {
            Function f = it.next();
            if (f.getName().equals(name) || f.getName(true).endsWith("::" + name)) {
                funcs.add(f);
            }
        }
        return funcs;
    }

    private List<String> getCalledNames(Function f) {
        ArrayList<String> names = new ArrayList<>();
        HashSet<String> seen = new HashSet<>();
        ReferenceManager rm = currentProgram.getReferenceManager();
        InstructionIterator it = currentProgram.getListing().getInstructions(f.getBody(), true);
        while (it.hasNext()) {
            Instruction ins = it.next();
            for (Reference ref : rm.getReferencesFrom(ins.getAddress())) {
                if (!ref.getReferenceType().isCall()) {
                    continue;
                }
                String name = null;
                Symbol s = getSymbolAt(ref.getToAddress());
                if (s != null) {
                    name = s.getName(true);
                }
                Function callee = getFunctionAt(ref.getToAddress());
                if (callee != null) {
                    name = callee.getName(true);
                }
                if (name == null) {
                    name = ref.getToAddress().toString();
                }
                if (!seen.contains(name)) {
                    seen.add(name);
                    names.add(name);
                }
            }
        }
        return names;
    }
}

File: tools/ghidra/legacy/NativeI2CStaticReport.py
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

patterns = [
    "I2C", "DDC", "NVAPI", "NvAPI", "Boost", "Intel", "ctlI2CAccess",
    "GvIntelI2C", "GvWriteI2C", "GvReadI2C", "nvapi_Direct", "Direct_Get"
]

targets = [
    "GvWriteI2C", "GvReadI2C", "GvInitDispLib", "GvFreeDispLib",
    "GvIsNvBoost5", "nvapi_Direct_Get", "nvapi_Direct_GetMethod"
]

def println(s=""):
    print(s)

def decompile_function(func, ifc):
    res = ifc.decompileFunction(func, 90, ConsoleTaskMonitor())
    if res and res.decompileCompleted():
        return res.getDecompiledFunction().getC()
    return "<decompile failed>"

def get_called_names(func):
    names = []
    body = func.getBody()
    refs = currentProgram.getReferenceManager()
    it = currentProgram.getListing().getInstructions(body, True)
    seen = set()
    for ins in it:
        for ref in refs.getReferencesFrom(ins.getAddress()):
            if ref.getReferenceType().isCall():
                sym = getSymbolAt(ref.getToAddress())
                if sym is not None:
                    name = sym.getName(True)
                else:
                    f = getFunctionAt(ref.getToAddress())
                    name = f.getName(True) if f is not None else str(ref.getToAddress())
                if name not in seen:
                    seen.add(name)
                    names.append(name)
    return names

def print_exports():
    println("## EXPORTS")
    syms = currentProgram.getSymbolTable().getSymbols("Exports")
    for s in syms:
        println("%s %s" % (s.getAddress(), s.getName(True)))
    println("")

def print_imports():
    println("## IMPORTS/EXTERNALS")
    ext = currentProgram.getExternalManager()
    for lib in ext.getExternalLibraryNames():
        println("LIB %s" % lib)
        for loc in ext.getExternalLocations(lib):
            println("  %s %s" % (loc.getAddress(), loc.getLabel()))
    println("")

def print_strings():
    println("## MATCHED STRINGS")
    listing = currentProgram.getListing()
    data = listing.getDefinedData(True)
    for d in data:
        if not d.hasStringValue():
            continue
        try:
            val = str(d.getValue())
        except:
            continue
        for p in patterns:
            if p.lower() in val.lower():
                println("%s %s" % (d.getAddress(), val))
                break
    println("")

def print_target_functions():
    ifc = DecompInterface()
    ifc.openProgram(currentProgram)
    fm = currentProgram.getFunctionManager()
    println("## TARGET FUNCTIONS")
    for name in targets:
        funcs = getGlobalFunctions(name)
        if not funcs:
            continue
        for func in funcs:
            println("### %s %s" % (func.getName(), func.getEntryPoint()))
            println("CALLS: %s" % ", ".join(get_called_names(func)))
            println(decompile_function(func, ifc))
            println("")
    println("")

println("# PROGRAM %s" % currentProgram.getName())
println("IMAGE_BASE %s" % currentProgram.getImageBase())
println("LANGUAGE %s" % currentProgram.getLanguageID())
println("COMPILER %s" % currentProgram.getCompilerSpec().getCompilerSpecID())
println("")
print_exports()
print_imports()
print_strings()
print_target_functions()
